composer/datasets/c4.py

```diff
@@ -0,0 +1,333 @@
+                f"so please be cautious to not request more tokens than the size of the original dataset.")
+
+        # Build tokenizer
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(tokenizer_name)
+        if self.tokenizer_name in ["gpt2"]:
+            # The GPT2 tokenizer has no padding token which causes bugs
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        # Load and shard dataset
+        text_dataset = datasets.load_dataset(path="c4", name="en", split=split, streaming=True)
+        text_dataset = self._shard_dataset(text_dataset)
+        if not isinstance(text_dataset, datasets.IterableDataset):
+            raise ValueError("Unable to build sharded Huggingface C4 Dataset.")
+
+        # Shuffle dataset
+        # Per-device shard list is shuffled at the start of each epoch
+        # Samples are read into and randomly sampled from per-device shuffle buffer
+        if self.shuffle:
+            text_dataset = text_dataset.shuffle(buffer_size=self.shuffle_buffer_size, seed=self.seed)
+
+        # Map text samples to token samples
+        # NOTE: Mapping is executed in batched mode for better CPU utilization,
+        # but the returned dataset is still an iterable over tokenized samples
+        text_sample_batch_size = 1000
+        token_dataset = text_dataset.map(
+            self._tokenize,
+            batched=True,
+            batch_size=text_sample_batch_size,
+        )
+
+        if self.group_method != "truncate":
+            # Map variable-length token samples to fixed-length token samples
+            # NOTE: Mapping is executed in batched mode for better CPU utilization,
+            # but the returned dataset is still an iterable over tokenized samples.
+            # NOTE: Depending on the 'group_method', this step may alter the number of
+            # token samples in the dataset, and may mix neighboring token samples together.
+            token_sample_batch_size = 1000
+            token_dataset = token_dataset.map(
+                self._group_tokens,
+                batched=True,
+                batch_size=token_sample_batch_size,
+            )
+
+        # Limit the number of post-processed samples
+        self.iterable_dataset = token_dataset.take(self.max_samples_per_device)
+
+    def __iter__(self):
+        worker_info = get_worker_info()
+        if worker_info is not None and worker_info.num_workers != 1:
+            raise ValueError("Multi-worker processing not supported for this dataset yet, please use 'num_workers=1'.")
+        return iter(self.iterable_dataset)
+
+    def __len__(self):
+        return self.max_samples_per_device
+
+    def _subsample(self, device_offset, text_batch):
+        # Only return the i-th item out of N sequential items
+        for k, v in text_batch.items():
+            text_batch[k] = v[device_offset:device_offset + 1]
+        return text_batch
+
+    def _shard_dataset(self, dataset):
```

Comment:
even those these are private functions, i think we should have good docstrings because:
1) any step of the data preprocessing is important for users to understand
2) users may want to extend this to other datasets themselves